### PR TITLE
Add osu!catch "simulate" command

### PR DESCRIPTION
**Pull request by infernalfire72**

### CODE REVIEW

**Comment by peppy** [comment]
The code quality of this is lacking a touch. Please go through the CI issues (hopefully you can run the build script locally to determine what the remaining issues are).

**Comment by infernalfire72** [comment]
fixed all issues (and checks are passing now)

File: PerformanceCalculator/Simulate/CatchSimulateCommand.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+                                                                                            + " Values: hr, dt, hd, fl, ez, etc...")]
+        public override string[] Mods { get; }
+
+        public override Ruleset Ruleset => new CatchRuleset();
+
+        protected override Dictionary<HitResult, int> GenerateHitResults(double accuracy, IBeatmap beatmap, int countMiss, int? countMeh, int? countGood) =>
+            new Dictionary<HitResult, int>
+            {
+                [HitResult.Perfect] = beatmap.HitObjects.OfType<Fruit>().Count() + beatmap.HitObjects.OfType<JuiceStream>().Sum(s => s.RepeatCount) + (beatmap.HitObjects.OfType<JuiceStream>().Count() * 2) /* - beatmap.HitObjects.OfType<BananaShower>().Sum(s => s.NestedHitObjects.Count - 1)*/,
+                [HitResult.Good] = beatmap.HitObjects.OfType<JuiceStream>().Sum(s => s.NestedHitObjects.OfType<Droplet>().Sum(d => (d is TinyDroplet) ? 0 : 1)), //beatmap.HitObjects.OfType<JuiceStream>().Sum(s => s.NestedHitObjects.Count() - s.NestedHitObjects.OfType<TinyDroplet>().Count() - s.RepeatCount - 2), // -2 because of slider start and end objects
```
Comment (peppy) [inline PerformanceCalculator/Simulate/CatchSimulateCommand.cs]: remove commented code. we generally don't allow this. replace with explanations or todo or something (i don't know why it's there but commented right now.

Comment (infernalfire72) [reply PerformanceCalculator/Simulate/CatchSimulateCommand.cs]: oh i just noted down previous approaches so i could reuse them if needed, going to remove these

File: PerformanceCalculator/Simulate/CatchSimulateCommand.cs [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    public class CatchSimulateCommand : SimulateCommand
+    {
+        [UsedImplicitly]
+        [Required, FileExists]
+        [Argument(0, Name = "beatmap", Description = "Required. The beatmap file (.osu).")]
+        public override string Beatmap { get; }
+
+        [UsedImplicitly]
+        [Option(Template = "-c|--combo <combo>", Description = "Maximum combo during play. Defaults to beatmap maximum.")]
+        public override int? Combo { get; }
```
Comment (smoogipoo) [inline PerformanceCalculator/Simulate/CatchSimulateCommand.cs]: Unused, requires implementation.

File: PerformanceCalculator/Simulate/CatchSimulateCommand.cs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+        [UsedImplicitly]
+        [Option(Template = "-C|--percent-combo <combo>", Description = "Percentage of beatmap maximum combo achieved. Alternative to combo option."
+                                                                       + " Enter as decimal 0-100.")]
+        public override double PercentCombo { get; } = 100;
```
Comment (smoogipoo) [inline PerformanceCalculator/Simulate/CatchSimulateCommand.cs]: Unused, requires implementation.

File: PerformanceCalculator/Simulate/CatchSimulateCommand.cs [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        public override string[] Mods { get; }
+
+        public override Ruleset Ruleset => new CatchRuleset();
+
+        protected override Dictionary<HitResult, int> GenerateHitResults(double accuracy, IBeatmap beatmap, int countMiss, int? countMeh, int? countGood) =>
+            new Dictionary<HitResult, int>
+            {
+                [HitResult.Perfect] = beatmap.HitObjects.OfType<Fruit>().Count() + beatmap.HitObjects.OfType<JuiceStream>().Sum(s => s.RepeatCount) + (beatmap.HitObjects.OfType<JuiceStream>().Count() * 2),
+                [HitResult.Good] = beatmap.HitObjects.OfType<JuiceStream>().Sum(s => s.NestedHitObjects.OfType<Droplet>().Sum(d => (d is TinyDroplet) ? 0 : 1)),
+                [HitResult.Meh] = beatmap.HitObjects.OfType<JuiceStream>().Sum(s => s.NestedHitObjects.OfType<TinyDroplet>().Count()),
```
Comment (smoogipoo) [inline PerformanceCalculator/Simulate/CatchSimulateCommand.cs]: This implementation doesn't look quite right. Check how taiko does it for reference: https://github.com/ppy/osu-tools/blob/89bf3cf6ff9a2d25d9a25d296d2e11ee41d55f28/PerformanceCalculator/Simulate/TaikoSimulateCommand.cs#L60-L84

**Comment by Joehuu** [comment]
Already added in https://github.com/ppy/osu-tools/pull/50, can be closed.

**Pull request closed without merging**